src/lib.rs

```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
-    spanned::Spanned,
-    token,
+	ext::IdentExt,
+	parse::{Parse, ParseStream},
+	parse_macro_input, parse_quote,
+	spanned::Spanned,
+	token,
 };
 
 #[proc_macro_attribute]
 pub fn sealed(args: TokenStream, input: TokenStream) -> TokenStream {
-    match parse_macro_input!(input) {
-        syn::Item::Impl(item_impl) => parse_sealed_impl(&item_impl),
-        syn::Item::Trait(item_trait) => {
-            Ok(parse_sealed_trait(item_trait, parse_macro_input!(args)))
-        }
-        _ => Err(syn::Error::new(Span::call_site(), "expected impl or trait")),
-    }
-    .unwrap_or_else(|e| e.to_compile_error())
-    .into()
+	match parse_macro_input!(input) {
+		syn::Item::Impl(mut item_impl) => parse_sealed_impl(&mut item_impl),
+		syn::Item::Trait(item_trait) => {
+			Ok(parse_sealed_trait(item_trait, parse_macro_input!(args)))
+		}
+		_ => Err(syn::Error::new(Span::call_site(), "expected impl or trait")),
+	}
+	.unwrap_or_else(|e| e.to_compile_error())
+	.into()
 }
 
 // Care for https://gist.github.com/Koxiaet/8c05ebd4e0e9347eb05f265dfb7252e1#procedural-macros-support-renaming-the-crate
 fn parse_sealed_trait(mut item_trait: syn::ItemTrait, args: TraitArguments) -> TokenStream2 {
-    let trait_ident = &item_trait.ident.unraw();
-    let trait_generics = &item_trait.generics;
-    let seal = seal_name(trait_ident);
-    let vis = &args.visibility;
-
-    let (_, ty_generics, where_clause) = trait_generics.split_for_impl();
-    item_trait
-        .supertraits
-        .push(parse_quote!( #seal::Sealed #ty_generics ));
-
-    let mod_code = if args.erased {
-        let lifetimes = trait_generics.lifetimes();
-        let const_params = trait_generics.const_params();
-        let type_params =
-            trait_generics
-                .type_params()
-                .map(|syn::TypeParam { ident, .. }| -> syn::TypeParam {
-                    parse_quote!( #ident : ?Sized )
-                });
-
-        quote! {
-            pub trait Sealed< #(#lifetimes ,)* #(#type_params ,)* #(#const_params ,)* > {}
-        }
-    } else {
-        // `trait_generics` does not output its where clause when tokenized (due
-        // to supertraits in the middle). So we output them separately.
-        quote! {
-            use super::*;
-            pub trait Sealed #trait_generics #where_clause {}
-        }
-    };
-
-    quote! {
-        #[automatically_derived]
-        #vis mod #seal {
-            #mod_code
-        }
-        #item_trait
-    }
+	let trait_ident = &item_trait.ident.unraw();
+	let trait_generics = &item_trait.generics;
+	let seal = seal_name(trait_ident);
+	let vis = &args.visibility;
+
+	let (_, ty_generics, where_clause) = trait_generics.split_for_impl();
+	item_trait
+		.supertraits
+		.push(parse_quote!( #seal::Sealed #ty_generics ));
+
+	let mut wrappers = Vec::<syn::TraitItem>::new();
+
+	let mut all_implementable = true;
+	let mut none_implementable = true;
+
+	for item in &mut item_trait.items {
```

Comment:
The new partial behavior represents a different approach to sealing, thus all the new behavior should be in its separate function, guarded by an `if args.partial`, you can extract the existing code to a separate function too in order to simplify code generation in the end.

To be more explicit, the macro should generate everything as it was until this change, and if `partial` is on, generate something else.